lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -11,23 +11,42 @@
 from flask_restful import abort
 
 from pbench.server import PbenchServerConfig
+from pbench.server.api.auth import Auth
 from pbench.server.api.resources import (
     API_OPERATION,
     ApiBase,
     JSON,
     PostprocessError,
     Schema,
     UnauthorizedAccess,
-    UnsupportedAccessMode,
 )
 from pbench.server.database.models.template import Template
 from pbench.server.database.models.datasets import Dataset
+from pbench.server.database.models.users import User
 
 # A type defined to allow the preprocess subclass method to provide shared
 # context with the assemble and postprocess methods.
 CONTEXT = Dict[str, Any]
 
 
+class NoSelectedDatasets(Exception):
+    """
+    Raised by the query builder when the input terms mean that no Elasticsearch
+    data could be selected without violating security rules. Most cases are
+    handled up-front by schema validation and authorization; edge cases raise
+    this exception to trigger logic in the subclass postprocess() method that
+    generates an appropriate empty response (e.g., [] or {}).
+    """
+
+    def __init__(self, user: str, access: str):
+        self.user = user
+        self.access = access
+
+    def __str__(self) -> str:
+        user = f"user {self.user!r}" if self.user else "unauthorized client"
+        return f"Query from {user} for access {self.access!r} cannot produce results"
```

Comment:
I certainly concur that the onus is on the Pbench server to avoid returning to the user items which s/he is not entitled to access, but I would effect that by ensuring that ES doesn't return such items to the server in the first place, by constructing the ES query in such a way as to prevent that.  And, I concur that, for the cases where we can predict a priori that the query will return no data (for any of several conceivable reasons), we can skip issuing the query to ES by raising this exception instead.

Thus, raising the exception indicates that, had we issued the query to ES, there would be no results to return to the user; it indicates that the query need not and should not be issued and, indeed, that the function has therefore declined to construct the query payload.

So, from the user-visibility perspective, the request "would not" produce results; from the server perspective, the query "must not" produce results and therefore "shall not", thanks to the exception.

Anyway, I think we've probably now given this more attention than it deserved, and I'm sure that it'll be fine, so I'm going to go ahead and mark this resolved.